clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -2339,6 +2345,28 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
   // LLVM counterparts if the call is marked 'const' (known to never set errno).
   // In case FP exceptions are enabled, the experimental versions of the
   // intrinsics model those.
+  bool ConstAlways =
+      getContext().BuiltinInfo.isConst(BuiltinID);
+
+  // There's a special case with the fma builtins where they are always const
+  // if the target environment is GNU or the target is OS is Windows and we're
+  // targeting the MSVCRT.dll environment.
+  switch (BuiltinID) {
```

Comment:
I worry about this list getting out of sync with reality again; it would be nice if we had some way to get this information out of `BuiltinInfo` instead of having a separate list here. If that's a difficult change to make, I don't insist on it happening now, but then we should add a FIXME comment to remember to come back and clean this up.

Reply:
hmm! First quick response would be, not sure where to get that from? But I will look at it longer to see if there is a way. @andykaylor WDYT?

Reply:
Can the special case in the math-errno fix concerning "fma" become actually a list of those? Aaron is worried that the list can become out of sync. We could get the name of the builtin via the getContext().BuiltinInfo.getName(BuiltinID); and work with that. Then we don't have to have a list here?

Reply:
@zahiraam I think the best way to do that would be to add a new letter vcode to the attributes here: https://github.com/llvm/llvm-project/blob/cbdccb30c23f71f20d05b19256232419e7c5e517/clang/include/clang/Basic/Builtins.def#L74 -- something like "//  m -> const when we GNU or MSVCRT libraries are targeted" and then fma, for example, would become "BUILTIN(__builtin_fma, "dddd", "Fmne")"

I'd suggest doing that as a separate patch, because it isn't obvious exactly how it should be managed.

I'm more concerned about the fact that Sema is marking calls as const in the AST based on the builtins information but we later discover that they aren't, in fact, const because of a pragma being used. @AaronBallman do you have any ideas about how that could be improved?

Reply:
> @zahiraam I think the best way to do that would be to add a new letter vcode to the attributes here:
> 
> https://github.com/llvm/llvm-project/blob/cbdccb30c23f71f20d05b19256232419e7c5e517/clang/include/clang/Basic/Builtins.def#L74
> 
> -- something like "// m -> const when we GNU or MSVCRT libraries are targeted" and then fma, for example, would become "BUILTIN(__builtin_fma, "dddd", "Fmne")"
> I'd suggest doing that as a separate patch, because it isn't obvious exactly how it should be managed.
> 


Yes, I can work on that in a subsequent patch. Will keep the FIXME comment then.